### PR TITLE
  perf(cors): pre-join static array header options during initialization

### PR DESCRIPTION
**Pull request by spellsaif**
Pre-joins static array options (`exposeHeaders`, `allowMethods`, and `allowHeaders`) once during `cors(options)` middleware initialization to eliminate per-request `.join(',')` string allocations.

### Motivation
In `src/middleware/cors/index.ts`, static array options such as `opts.exposeHeaders` were joined via `.join(',')` on every single request, and `allowMethods` / `allowHeaders` were joined on every preflight `OPTIONS` request.

Since these configuration arrays are static for a given CORS middleware instance, calling `Array.prototype.join(',')` on every request creates unnecessary heap string allocations.

### Solution
Pre-join static `exposeHeaders`, `allowMethods`, and `allowHeaders` arrays into immutable strings once during middleware setup.
Passing pre-joined string references into response header setters avoids per-request string joining.

### Testing
- `npm test src/middleware/cors/index.test.ts` passes 100% cleanly (23/23 tests passing with 98% coverage).
- All 1,173 repository unit tests pass.

### The author should do the following, if applicable

- [ ] Add tests
- [x] Run tests
- [x] `bun run format:fix && bun run lint:fix` to format the code
- [x] Add [TSDoc](https://tsdoc.org/)/[JSDoc](https://jsdoc.app/about-getting-started) to document the code

### CODE REVIEW

**Comment by usualoma** [comment]
Hi @spellsaif,
Thanks for opening this PR.

Pre-joining the static array options during middleware initialization looks like a reasonable optimization.

I have one concern about `Access-Control-Allow-Methods`. When an `allowMethods` function returns an empty array, the current implementation omits the header, while this PR emits `Access-Control-Allow-Methods` with an empty value because an empty array is truthy.

A missing header and a present-but-empty header have different CORS security semantics. In the Fetch preflight algorithm, a missing value can receive special fallback handling, whereas an empty value is parsed as an explicit empty method list:
https://fetch.spec.whatwg.org/#cors-preflight-fetch

Since this PR is intended as a performance optimization, I think it should preserve the existing semantics. Could you normalize the function result to a string inside `findAllowMethods`, for example:

```diff
diff --git a/src/middleware/cors/index.test.ts b/src/middleware/cors/index.test.ts
index 9ea184c7..ee9efdf2 100644
--- a/src/middleware/cors/index.test.ts
+++ b/src/middleware/cors/index.test.ts
@@ -382,6 +382,17 @@ describe('CORS by Middleware', () => {
     expect(res2.headers.get('Access-Control-Allow-Methods')).toBe('GET,HEAD')
   })
 
+  it('Does not set allow methods when function returns an empty array', async () => {
+    for (const allowMethods of [() => [], async () => []]) {
+      const app = new Hono()
+      app.use(cors({ allowMethods }))
+
+      const res = await app.request('http://localhost/', { method: 'OPTIONS' })
+
+      expect(res.headers.get('Access-Control-Allow-Methods')).toBeNull()
+    }
+  })
+
   it('Emits the wildcard, not the reflected origin, when credentials is true with wildcard origin', async () => {
     const res = await app.request('http://localhost/api10/abc', {
       headers: {
diff --git a/src/middleware/cors/index.ts b/src/middleware/cors/index.ts
index 4febd409..1ea2e012 100644
--- a/src/middleware/cors/index.ts
+++ b/src/middleware/cors/index.ts
@@ -88,7 +88,7 @@ export const cors = (options?: CORSOptions): MiddlewareHandler => {
 
   const findAllowMethods = ((optsAllowMethods) => {
     if (typeof optsAllowMethods === 'function') {
-      return optsAllowMethods
+      return async (origin: string, c: Context) => (await optsAllowMethods(origin, c)).join(',')
     } else if (Array.isArray(optsAllowMethods)) {
       const methodsStr = optsAllowMethods.join(',')
       return () => methodsStr
@@ -126,10 +126,7 @@ export const cors = (options?: CORSOptions): MiddlewareHandler => {
 
       const allowMethods = await findAllowMethods(c.req.header('origin') || '', c)
       if (allowMethods) {
-        set(
-          'Access-Control-Allow-Methods',
-          typeof allowMethods === 'string' ? allowMethods : allowMethods.join(',')
-        )
+        set('Access-Control-Allow-Methods', allowMethods)
       }
 
       let headersStr = allowHeadersStr
```

**Comment by spellsaif** [comment]
Hi @usualoma ,

Thank you for catching this edge case! You're completely right about the distinction in CORS preflight semantics when `Access-Control-Allow-Methods` is omitted vs. present with an empty value.

I've updated the PR to normalize the return value of `allowMethods` inside `findAllowMethods` to a string as suggested, while preserving the pre-joining optimization for static array options. I've also added the test case covering empty array returns for both synchronous and asynchronous functions.

**Comment by usualoma** [comment]
Hi @spellsaif,
Thank you. I think it's okay to merge it as is.

**Review by yusukebe — APPROVED** [review]
LGTM!

**Comment by yusukebe** [comment]
> Pre-joins static array options (`exposeHeaders`, `allowMethods`, and `allowHeaders`) once during `cors(options)` middleware initialization to eliminate per-request `.join(',')` string allocations.

Ah, I see. Doing things in the initialization phase is a good performance improvement. Thanks!
